File: 112.PathSum/Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }

        return calSum(root, 0, sum);
    }


    public boolean calSum(TreeNode node, int pre_sum, int target) {

        int sum = pre_sum + node.val;

        if (node.left == null && node.right == null) {
            if (sum == target) {
                return true;
            }
            return false;
        }

        boolean flag = false;

        if (node.left != null) {
            flag = calSum(node.left, sum, target);
        }

        if (flag == false && node.right != null) {
            flag = calSum(node.right, sum, target);
        }

        return flag;
    }


    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null || (root.left == null && root.right == null)) {
            return root != null && root.val == sum;
        }

        return hasPathSum(root.left, sum - root.val) || hasPathSum(root.right, sum - root.val);
    }
}

File: 112.PathSum/sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if not root:
            return False

        return self.calSum(root, 0, sum)


    def calSum(self, node, pre_sum, target):
        sum = pre_sum + node.val

        if not node.left and not node.right and sum == target:
            return True

        flag = False

        if node.left:
            flag = self.calSum(node.left, sum, target)
        if node.right and flag == False:
            flag = self.calSum(node.right, sum, target)

        return flag


    def hasPathSum(self, root, sum):
        if not root or (not root.left and not root.right):
            return root is not None and root.val == sum

        return self.hasPathSum(root.left, sum - root.val) or self.hasPathSum(root.right, sum - root.val)
